### DISCUSSION TITLE
Can I install arm64 libraries on x86_64 with homebrew?

### DISCUSSION THREAD

**Discussion by maxirmx** [Homebrew/discussions#2843] [Getting started] [answered]
I am cross-compiling on x86_64 MacOS 11 for arm64 architecture.
clang/XCode support it but I face an issue when external library is required.
Let it be boost, for example.  I know that the bottle for arm64 is available but it looks like there is no way to select it for installation.

`arch -arm64 brew ... `   says that this architecture is unknown which sounds fair.

So the question is wheather there is an option to force brew install bottles for 'foreign' architecture ?

**Comment by carlocab** [comment, marked as answer]
Not easily, since this isn't really a supported use-case.

What you can do is something like
```
brew fetch --force --bottle-tag=arm64_big_sur <formula>
brew install /path/to/downloaded_bottle.tar.gz
```
This won't work if you already have the x86_64 version of the formula installed, though.

**Reply by maxirmx** [reply]
Thank you, I made it work with something like this:

```
setopt sh_word_split
mkdir arm-homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C arm-homebrew 
alias arm-brew='$(pwd)/arm-homebrew/bin/brew'
response=$(arm-brew fetch --force --bottle-tag=arm64_big_sur boost | grep "Downloaded to")
parsed=($response)  
arm-brew install $parsed[3] 
```

**Reply by reuben** [reply]
This will install the x86_64 version of any dependencies declared in the bottle :( In my case I'm trying to cross compile with a dependency on sox, which has its own transitive dependencies, and I need to be able to link against the arm64 version of the entire dependency tree.

**Reply by gromgit** [reply]
```
$ man brew
[...]
  fetch [options] formula|cask [...]
       Download a bottle (if available) or source packages for formulae and binaries for casks. For files, also print SHA-256 checksums.
[...]
       --deps Also download dependencies for any listed formula.
```
So try:
```
brew fetch --force --deps --bottle-tag=arm64_big_sur sox
for b in <list_of_downloaded_bottle_paths>; do
  brew install "$b"
done
```

**Reply by carlocab** [reply]
Try using `brew deps -n`. See `brew help deps` to see what this does.

Or, patch `brew` to trick it into thinking you're on an M1 machine.

**Reply by maxirmx** [reply]
> This will install the x86_64 version of any dependencies declared in the bottle :( In my case I'm trying to cross compile with a dependency on sox, which has its own transitive dependencies, and I need to be able to link against the arm64 version of the entire dependency tree.

I would advise that you first hardcode the whole list of depndencies including transitive ones and see how far you can go this way.

Just in case you are able to find something useful in my scripts
https://github.com/tamatebako/dwarfs/tree/libdwarfs_wr_branch
https://github.com/tamatebako/libdwarfs/tree/maxirmx_v0.1.9

**Reply by reuben** [reply]
I tried the scripting approach with a separate Homebrew checkout but it doesn't work for some bottles:

```
$ arm-brew fetch --deps --force --bottle-tag=arm64_big_sur sox | grep "Downloaded to:" | awk '{ print $3 }' | xargs -n 1 arm-brew install --ignore-dependencies
Warning: `--ignore-dependencies` is an unsupported Homebrew developer flag!
Adjust your PATH to put any preferred versions of applications earlier in the
PATH rather than using this unsupported flag!

==> Pouring sox--14.4.2_3.arm64_big_sur.bottle.tar.gz
🍺  /Users/reubenmorais/Development/arm-homebrew/Cellar/sox/14.4.2_3: 25 files, 2.0MB
==> Running `brew cleanup sox`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Warning: `--ignore-dependencies` is an unsupported Homebrew developer flag!
Adjust your PATH to put any preferred versions of applications earlier in the
PATH rather than using this unsupported flag!

Warning: Building pkg-config from source:
  The bottle only works in the default /usr/local location.
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz
############                                                              17.1%^C
One sec, cleaning up...
```

**Reply by reuben** [reply]
Ah, looks like `--force-bottle` will do it :D

**Reply by reuben** [reply]
Full solution:

```bash
mkdir -p ~/arm-target/bin
mkdir -p ~/arm-target/brew-cache
export PATH="$HOME/arm-target/bin:$PATH"

cd ~/arm-target
mkdir arm-homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C arm-homebrew
ln -s ~/arm-target/arm-homebrew/bin/brew ~/arm-target/bin/arm-brew

export HOMEBREW_CACHE=~/arm-target/brew-cache
export HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK=1
arm-brew fetch --deps --bottle-tag=arm64_big_sur sox opusfile |\
  grep -E "(Downloaded to:|Already downloaded:)" |\
  grep -v pkg-config |\
  awk '{ print $3 }' |\
  xargs -n 1 arm-brew install --ignore-dependencies --force-bottle

# Install host version of pkg-config so we can call it in the build system
arm-brew install pkg-config
ln -s ~/arm-target/arm-homebrew/bin/pkg-config ~/arm-target/bin/arm-pkg-config
```

**Reply by holliebeth76** [reply]
Make.sure everything is complete

**Reply by alexcohn** [reply]
You can use `$(brew --cache --bottle-tag=arm64_big_sur ${FORMULA})` instead of parsing the output of **fetch**.

**Reply by DBraun** [reply]
@alexcohn Can you share your whole code which worked?

**Reply by alexcohn** [reply]
@DBraun I don't have "code" per se. And my case is just the opposite: I wanted x86_64 libraries for cross-compiling something on an Apple Silicon machine.

`brew --cache --bottle-tag=arm64_big_sur ${FORMULA}` produces the path of a **.tar.gz** that you need for `brew install` as described by @carlocab above.

**Reply by DBraun** [reply]
Ok I'm trying this on GitHub actions with a macos-12 runner (note the `monterey`):

```bash
PACKAGES=(gtk+3 qt@5 libsndfile libmicrohttpd)
for PACKAGE in "${PACKAGES[@]}"
do
    result=$(brew --cache --bottle-tag=arm64_monterey $PACKAGE)
    echo "Looking for arm64 version of $result"
    brew install $result
done
```
But I get errors:
```
Looking for arm64 version of /Users/runner/Library/Caches/Homebrew/downloads/a5ed68a7ed25fc8166b9187dfe38edfd30149c32d3a719afc9dee9f927421dda--gtk+3--3.24.37.arm64_monterey.bottle.tar.gz
Error: No such file or directory @ realpath_rec - /Users/runner/Library/Caches/Homebrew/downloads/a5ed68a7ed25fc8166b9187dfe38edfd30149c32d3a719afc9dee9f927421dda--gtk+3--3.24.37.arm64_monterey.bottle.tar.gz
Error: Process completed with exit code 1.
```

I tried the same with macos-11 and `big_sur` but got a similar error. It works fine on my arm64 laptop.

**Reply by alexcohn** [reply]
@DBraun you seem to have missed the `brew fetch` step. My improvement only gets rid of parsing of output of `brew fetch`, but does not replace it.

**Comment by maxirmx** [comment]
@DBraun 
I have it running at https://github.com/tamatebako/dwarfs/blob/main/.github/workflows/macos-arm64.yml and https://github.com/tamatebako/libdwarfs/blob/main/.github/workflows/macos-arm64.yml

(some two-way scripts here https://github.com/tamatebako/tebako-tools/tree/master/ci-scripts)

I didn't have time to make it as neat as the commenters suggested, but it works

There are more suggestions here: https://stackoverflow.com/questions/70821136/can-i-install-arm64-libraries-on-x86-64-with-homebrew/70822921#70822921
